runtime/compiler/control/HookedByTheJit.cpp

```diff
@@ -6004,6 +6008,52 @@ static void suspendSamplerThreadForCheckpoint(J9VMThread *samplerThread, J9JITCo
       }
    compInfo->releaseCompMonitor(samplerThread);
    }
+
+static void triggerRecompilationForPreCheckpointGeneratedFSDBodies(J9VMThread *samplerThread, TR_J9VMBase *fe, TR::CompilationInfo *compInfo)
+   {
+   compInfo->acquireCompMonitor(samplerThread);
+
+   J9Method *method;
+   while (method = compInfo->popForcedRecompilation())
+      {
+      J9ROMMethod *romMethod = J9_ROM_METHOD_FROM_RAM_METHOD(method);
+      if (!_J9ROMMETHOD_J9MODIFIER_IS_SET(romMethod, J9AccAbstract)
+            && !_J9ROMMETHOD_J9MODIFIER_IS_SET(romMethod, J9AccNative)
+#if !defined(J9VM_OPT_OPENJDK_METHODHANDLE)
+            && !_J9ROMMETHOD_J9MODIFIER_IS_SET(romMethod, J9AccMethodFrameIteratorSkip)
+#endif
+            && compInfo->isCompiled(method))
+         {
+         if (TR::Options::getCmdLineOptions()->getVerboseOption(TR_VerboseCheckpointRestoreDetails))
+            TR_VerboseLog::writeLineLocked(TR_Vlog_CHECKPOINT_RESTORE, "Attempting to force %p for recompilation", method);
+
+         TR_MethodEvent event;
+         event._eventType = TR_MethodEvent::ForcedRecompilationPostRestore;
+         event._j9method = method;
+         event._oldStartPC = method->extra;
+         event._vmThread = samplerThread;
+         event._classNeedingThunk = 0;
+         bool newPlanCreated = false;
+         TR_OptimizationPlan *plan = TR::CompilationController::getCompilationStrategy()->processEvent(&event, &newPlanCreated);
+         // the plan needs to be created only when async compilation is possible
+         // Otherwise the compilation will be triggered on next invocation
+         if (plan)
+            {
+            TR_PersistentJittedBodyInfo *bodyInfo = TR::Recompilation::getJittedBodyInfoFromPC(method->extra);
```

Comment:
If the method cannot be recompiled, the bodyInfo/methodInfo will be NULL. Do we guarantee that bodyInfo will exist here? I guess we should make sure when we add methods to this recompilation queue.

Reply:
Yeah that's a good point; I can make sure to only add methods if the bodyInfo exists.